test/dash-playlist-loader.test.js

```diff
@@ -30,51 +30,149 @@ QUnit.test('throws if the playlist url is empty or undefined', function(assert)
 
-  let newLoader =
-    new DashPlaylistLoader({ uri: 'playlist-1' }, this.fakeHls, false, loader);
+  assert.notOk(loader.started, 'not started');
 
-  assert.equal(newLoader.state, 'HAVE_METADATA', 'has metadata');
-  assert.ok(newLoader.started, 'started');
+  loader.load();
+  assert.equal(loader.state, 'HAVE_NOTHING', 'no metadata has loaded yet');
+  assert.ok(loader.started, 'started');
 });
 
-QUnit.test('requests the manifest immediately when given a URL', function(assert) {
+QUnit.skip('moves to HAVE_MASTER after loading a master playlist', function(assert) {
   let loader = new DashPlaylistLoader('dash.mpd', this.fakeHls);
+  let state;
+
+  loader.on('loadedplaylist', function() {
+    state = loader.state;
+  });
 
   loader.load();
-  assert.equal(this.requests.length, 1, 'made a request');
-  assert.equal(this.requests[0].url, 'dash.mpd', 'requested the manifest');
+  standardXHRResponse(this.requests.shift());
+
+  assert.ok(loader.master, 'the master playlist is available');
+  assert.strictEqual(state, 'HAVE_MASTER', 'the state at loadedplaylist correct');
 });
 
-QUnit.test('moves to HAVE_MASTER and HAVE_METADATA after loading the manifest',
-function(assert) {
+QUnit.test('moves to HAVE_METADATA after loading a media playlist', function(assert) {
+  let loadedPlaylist = 0;
+  let loadedMetadata = 0;
   let loader = new DashPlaylistLoader('dash.mpd', this.fakeHls);
-  let loadedPlaylistStates = [];
 
   loader.load();
 
   loader.on('loadedplaylist', function() {
-    loadedPlaylistStates.push(loader.state);
+    loadedPlaylist++;
+  });
+  loader.on('loadedmetadata', function() {
+    loadedMetadata++;
   });
+
   standardXHRResponse(this.requests.shift());
-  assert.ok(loader.master, 'the master playlist is available');
+  this.clock.tick(1);
+
   // because DASH only has one manifest, it should go through two loadedplaylists
   // and end with HAVE_METADATA because it already has the first media ready
-  assert.equal(loadedPlaylistStates.length, 2, 'triggered two loadedplaylist events');
-  assert.equal(loadedPlaylistStates[0], 'HAVE_MASTER', 'got master first');
-  assert.equal(loadedPlaylistStates[1], 'HAVE_METADATA', 'got media second');
+  assert.strictEqual(loadedPlaylist, 2,
+    'fired loadedplaylist once for master playlist and once for' +
+    ' autoselected first child playlist');
+  // this is triggered on the next tick after construction
+  assert.strictEqual(loadedMetadata, 1, 'fired loadedmetadata once');
+  assert.ok(loader.master, 'sets the master playlist');
+  assert.ok(loader.media(), 'sets the media playlist');
+  assert.strictEqual(this.requests.length, 0, 'no further requests are needed');
+
+  const childLoader = new DashPlaylistLoader(loader.master.playlists[0], this.fakeHls, false, loader);
+
+  childLoader.on('loadedplaylist', function() {
+    loadedPlaylist++;
+  });
+  childLoader.on('loadedmetadata', function() {
+    loadedMetadata++;
+  });
+
+  childLoader.load();
```

Comment:
Since the child loader will be "loading" a media playlist and not a master playlist, it *will* skip the `HAVE_MASTER` state and move directly to `HAVE_METADATA`, however *NOT* on construction, but after a call to `childLoader.load()`, which is *asynchronous*. This would result in a single `loadedplaylist` event and a single `loadedmetadata` event.
We should add assertions both before and after the `load` call to ensure this behavior is followed.
Before the call to load, no events should have been fired and the state should not have changed
```
loadedPlaylist === 2
loadedMetadata === 1
childLoader.state === HAVE_NOTHING
childLoader.media() === undefined
```
After the call to `load`
```
loadedPlaylist === 3
loadedMetadata === 2
childLoader.state === HAVE_METADATA
childLoader.media() !=== undefined
```

Reply:
After double checking the PlaylistLoader path, it does appear that the child loader does actually go through the `HAVE_MASTER` state briefly in `start()` when the master playlist is inferred. I plan to do the same in DashPlaylistLoader to make both paths be in line with each other